k2/csrc/ragged_ops.cu

```diff
@@ -1871,4 +1871,144 @@ Array1<int32_t> CoveringShapeForwardMap(RaggedShape &src,
   return ans;
 }
 
+
+
+void RaggedShapeAxis0Splitter::Init(RaggedShape &src) {
+  int32_t num_layers = src.NumLayers(),
+    num_layers_out = num_layers - 1,
+              dim0 = src.Dim0();
+  K2_CHECK_LE(num_layers_out, 4);  // If this fails, add something to the 4s and
+                                   // 5s here and in the header.
+  K2_CHECK_GT(num_layers, 1);
+
+  ContextPtr c = src.Context();
+  composite_row_splits_ = Array2<int32_t>(c, num_layers + 1, dim0 + 1);
+  Array2Accessor<int32_t> composite_row_splits_acc =
+      composite_row_splits_.Accessor();
+
+  RowSplitsAccessor<5> src_row_splits_acc(src);
+
+  ArrayAccessor<int32_t*, 5> row_splits_out_acc;
+  Array1<int32_t> garbage1(c, dim0 + dim0 + 1);  // won't be read.
+  row_splits_out_acc.data[0] = garbage1.Data();
+  for (int32_t l = 0; l < num_layers_out; l++) {
+    row_splits_out_[l] = Array1<int32_t>(c, src.TotSize(l + 1) + dim0 + 1);
+    row_splits_out_acc.data[l + 1] = row_splits_out_[l].Data();
+  }
+
+  // set composite_row_splits_ and also those elements of
+  // the output row_splits which are bound to be zero.
+  K2_EVAL(c, dim0 + 1, lambda_set_composite_row_splits, (int32_t i) -> void {
+      int32_t cur_pos = i;
+      composite_row_splits_acc(0, i) = cur_pos;
+      for (int32_t l = 0; l < num_layers; l++) {
+        // The following statement sets the zero at the beginning of each
+        // row_splits, plus a final zero that we write to avoid an if-statement.
+        row_splits_out_acc.data[l][cur_pos + i] = 0;
+        cur_pos = src_row_splits_acc.ptrs[l][cur_pos];
```

Comment:
`RowSplitsAccessor::operator()` is never used. Should we remove it? Seems it is good enough to access its `ptrs` directly.

Reply:
would rather keep for now in case someone wants to access it with the same ordering as RowSplits(1) etc.